Add "Risk review" and "Blocked" tabs to transactions page

eduardoumpierre opened the pull request:
Fixes Automattic/woocommerce-payments-server#2787

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

This PR aims to add two new tabs to the WCPay transactions page: "Risk review" and "Blocked". The new tabs contain the following features:
- Search by order id and customer name
- Sorting
- Pagination
- Download CSV

The UI is behind a feature flag.

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**woocommerce-payments-server**
* Apply the following diff:
```diff
diff --git a/server/wp-content/rest-api-plugins/endpoints/wcpay/class-intentions-controller.php b/server/wp-content/rest-api-plugins/endpoints/wcpay/class-intentions-controller.php
index 1a0b5276d..2103a7523 100644
--- a/server/wp-content/rest-api-plugins/endpoints/wcpay/class-intentions-controller.php
+++ b/server/wp-content/rest-api-plugins/endpoints/wcpay/class-intentions-controller.php
@@ -753,7 +753,7 @@ class Intentions_Controller extends Base_Controller {
 		}
 
 		// @todo Retrieve merchant ruleset setting.
-		$ruleset = [];
+		$ruleset = [ true ];
 
 		// We only want to apply the rules if the client version it at least 5.7.0 and
 		// the merchant ruleset setting is available.

```

* Follow the testing instructions from `add/3079-store-fraud-outcomes` (https://github.com/Automattic/woocommerce-payments-server/pull/3115)

**Client**
* Checkout to this branch `update/2787-transactions-page`
* Run `npm start`
* Listen to Stripe webhooks
* Apply the following diff:
```diff
diff --git a/includes/class-wc-payments.php b/includes/class-wc-payments.php
index 41aefb0f..98b6b05f 100644
--- a/includes/class-wc-payments.php
+++ b/includes/class-wc-payments.php
@@ -566,6 +566,19 @@ class WC_Payments {
 
 		add_action( 'admin_enqueue_scripts', [ __CLASS__, 'enqueue_assets_script' ] );
 		add_action( 'wp_enqueue_scripts', [ __CLASS__, 'enqueue_assets_script' ] );
+
+		add_filter(
+			'wcpay_api_request_params',
+			function( $params, $api, $method ) {
+				if ( false !== strpos( $api, WC_Payments_API_Client::INTENTIONS_API ) && WC_Payments_API_Client::POST === $method ) {
+					$params['fraud_settings_enabled'] = 'true';
+				}
+
+				return $params;
+			},
+			10,
+			3
+		);
 	}
 
 	/**
diff --git a/woocommerce-payments.php b/woocommerce-payments.php
index 8f8c6036..e66c1e72 100644
--- a/woocommerce-payments.php
+++ b/woocommerce-payments.php
@@ -12,7 +12,7 @@
  * WC tested up to: 7.3.0
  * Requires at least: 5.9
  * Requires PHP: 7.0
- * Version: 5.6.1
+ * Version: 5.7.0
  *
  * @package WooCommerce\Payments
  */

```
* Perform some order (at least 2 of each for testing purposes)
  * **Blocked order**
    * > **Note**
       > (Use incognito mode and open a new browser instance for every blocked order)
    * Open your test shop
    * Add an item to the shopping cart, go to the checkout page
    * The "Last name" should be Block
    * Use the card 4242 4242 4242 4242

  * **On review order**
    * Open your test shop
    * Add an item to the shopping cart, go to the checkout page
    * The "Last name" should be Review
    * Use the card 4242 4242 4242 4242
* Make sure you have `wcpay_fraud_protection_settings_active` flag on your client disabled
  * If you don't have it set yet, run: `npm run wp option add wcpay_fraud_protection_settings_active 0` 
  * Otherwise, update it by running: `npm run wp option update wcpay_fraud_protection_settings_active 0`
* Go to **Payments > Transactions**
* You should see just 2 tabs: "Transactions" and "Uncaptured"
<img width="618" alt="image" src="https://user-images.githubusercontent.com/16882226/226800443-6aa56e64-6999-434f-967c-306cd45957cc.png">

* Enable the `wcpay_fraud_protection_settings_active` flag on your client
  * If you don't have it set yet, run: `npm run wp option add wcpay_fraud_protection_settings_active 1` 
  * Otherwise, update it by running: `npm run wp option update wcpay_fraud_protection_settings_active 1`
* Go to **Payments > Transactions**
* You should see 4 tabs: "Transactions", "Uncaptured", "Risk review", and "Blocked"
<img width="618" alt="image" src="https://user-images.githubusercontent.com/16882226/226800597-e687aa2b-580a-45bf-bc32-373ea7835034.png">

* Explore both "Risk review" and "Blocked" tabs
  * You can tweak the `per_page` query param in order to trick the pagination
    * You can change it to `2` in order to test the pagination
  * It is possible to search by order id and customer name
  * Try to sort columns
  * Click on the download button
  * Make sure the pagination works

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

eduardoumpierre committed:
Extract TableCard rows interface

eduardoumpierre committed:
Add risk review test mode notice

eduardoumpierre committed:
Add risk review tab on transactions page

eduardoumpierre committed:
Add blocked transactions tab to transactions page

eduardoumpierre committed:
Add transaction status chip component

eduardoumpierre committed:
Fix transaction lists summary data

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into update/2787-transactions-page

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into update/2787-transactions-page

eduardoumpierre committed:
Add endpoint and data structure to retrieve blocked transactions

eduardoumpierre committed:
Add endpoint and data structure to retrieve on review transactions

eduardoumpierre committed:
Fix UI for flagged transactions list

eduardoumpierre committed:
Remove payment intent dependency when retrieving hold transactions

eduardoumpierre committed:
Add method to build the order info from a given fraud outcome entry

eduardoumpierre committed:
Fix amount column on blocked transactions list

eduardoumpierre committed:
Fix transactions status mapping

eduardoumpierre committed:
Update transactions chip to match design

eduardoumpierre committed:
Remove risk level column from transactions list

eduardoumpierre committed:
Add summary for on review and blocked transactions

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into update/2787-transactions-page

eduardoumpierre committed:
Add fraud outcome transactions pagination

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into update/2787-transactions-page

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `5487` or branch name `update/2787-transactions-page` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=update/2787-transactions-page) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 0021deefa3c43136bfbea6b3328286b487715a3d
- Build time: 2023-03-24 15:44:43 UTC

_Note: the build is updated when a new commit is pushed to this PR._

eduardoumpierre committed:
Rollback get risk level changes

eduardoumpierre committed:
Add order filters to fraud outcome transactions list

eduardoumpierre committed:
Add search on fraud outcome transactions list

eduardoumpierre committed:
Add download button to fraud outcome transactions list

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into update/2787-transactions-page

eduardoumpierre committed:
Add changelog entry

eduardoumpierre committed:
Refactor fraud outcome list methods to use a request class

eduardoumpierre committed:
Add feature flag condition to transactions list

eduardoumpierre committed:
Fix transactions list unit tests

eduardoumpierre committed:
Fix psalm errors

eduardoumpierre committed:
Replace string with constant

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into update/2787-transactions-page

eduardoumpierre committed:
Update fraud outcomes REST endpoint to match server

tpaksu commented on client/transactions/fraud-protection/autocompleter.tsx:
```diff
@@ -0,0 +1,72 @@
+	getOptionLabel( option: CompletionOption, query: string ): JSX.Element {
+		const match = computeSuggestionMatch( option.label, query );
+
+		return (
```
I assume if there's no match, it'll display an empty element, right? How about we return if match exists? Like `return match ? ( ..span ) : null` ?

eduardoumpierre replied:
I tried this approach, but the behavior didn't change. It's still rendering an empty item. I think the issue is that the `getOptionLabel` only formats the text. ([Reference](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/components/src/search/index.tsx#L234))

<img width="1172" alt="image" src="https://user-images.githubusercontent.com/16882226/226966876-d1710f7a-b2b2-42f2-a3f1-97457fe3e4d8.png">

tpaksu replied:
Maybe you can add a filter to the response of `apiFetch` to eliminate unmatching options?

eduardoumpierre replied:
We already have the filtering on the server, but I can definitely add an extra check to be safer!

eduardoumpierre replied:
Added on [626427e](https://github.com/Automattic/woocommerce-payments/pull/5487/commits/626427e681b18fee2e47a87bf8d00e69812dab44).

eduardoumpierre committed:
Add request documentation

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into update/2787-transactions-page

eduardoumpierre committed:
Add transaction status chip unit tests

eduardoumpierre committed:
Add fraud outcome transactions data unit tests

eduardoumpierre committed:
Add additional unit tests

eduardoumpierre committed:
Add fallback to fraud outcome transactions

It should fallback in case of the wcpay_fraud_outcome_not_found error.
It fallbacks the transactions list to an empty array, and
the transactions summary to an object with count and total set as 0.

eduardoumpierre committed:
Add match filter to fraud protection transactions autocompleter

eduardoumpierre committed:
Update fraud outcome transactions order search logic

tpaksu reviewed (approved):
Read the code (took me a while to complete it :D) and all is good, thanks for fixing the autocompleter thing btw! Also manual testing went well, didn't get any problems with all, including downloading the table as CSV. We can merge it! Thanks for working on this, LGTM!

allie500 committed:
Merge branch 'develop' into update/2787-transactions-page

eduardoumpierre committed:
Merge branch 'develop' into update/2787-transactions-page
